### PR TITLE
fix dtype mismatch when using vocab_tiling

### PR DESCRIPTION
**Pull request by aireenmei**
# Description
Encountered [this error](https://paste.googleplex.com/6595822797062144) when testing gemma4 with bfloat16 and `num_vocab_tiling=8`.

The root cause is the incompatibility of weight_dtype=bfloat16 and vocab_tiling. In my test, I was using `weight_dtype=float32 dtype=bfloat16`,  but gemma4 has a bug that uses config.dtype (should use config.weight_dtype) to initiate layer scalar, resulting in a bfloat16 weight that causes this error.

This PR fix:
* add dtype cast to allow using weight_dtype=bfloat16 with vocab_tiling. Although weight_dtype=bfloat16 is not common, I see some test scripts use it, so it's still good to support
* gemma4 layer scalar initiate with config.weight_dtype

# Tests
* Added unit test and verified that test fail without the fix and pass after.
* [test script](https://paste.googleplex.com/5252416829259776) 
* [log with error](https://cloudlogging.app.goo.gl/R3YseuHvduWy6cbe6)
* [log after fix](https://cloudlogging.app.goo.gl/r46gq5pKdcZmaoYV8)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/vocabulary_tiling.py [hunk 1]
```diff
@@ -228,6 +228,8 @@ def _bwd_scan_body(grad_params_acc, chunk_data):
     # grad_params = jax.tree_util.tree_map(lambda x, y: y.astype(x.dtype), gathered_params, grad_params)
```
Comment (NuojCheng) [inline src/maxtext/utils/vocabulary_tiling.py]: could you remove the comments in line 227 and 228?

**Review by NuojCheng — APPROVED** [review]
thank you for the fix!

**Review by gobbleturk — COMMENTED** [review]
Is there a compile test using vocab tiling that should have caught this? Can we add some test protecting this?

File: src/maxtext/utils/vocabulary_tiling.py (hunk 1 above)
Comment (aireenmei) [reply src/maxtext/utils/vocabulary_tiling.py]: Cleaned the comments as discussed offline

**Comment by aireenmei** [comment]
> Is there a compile test using vocab tiling that should have caught this? Can we add some test protecting this?

Added and updated PR decription

**Review by gobbleturk — APPROVED** [review]
Awesome, thank you for the test!
